### PR TITLE
perf(array-virtual-repeat-strategy): optimise case where mutations do…

### PR DESCRIPTION
**Pull request by lsesol**
…nt affect the array length and buffer sizes

In our use case, most of the mutation to the observed collection is replacing chunks of the underlying array. Essentially I create a fixed sized set of items (known number of results), then lazy load real data into it from remote source based up the current/predicted scroll position.

This change is simple: if and only if mutations only replace items in the array (splice.addedCount===splice.removed.length), then we can assume the top and bottom buffers don't require changes and we can simplify to only update view slots which are currently visible.

I think this is fairly innocuous and seems to work well in our application. However, I'm not yet completely up to scratch with Aurelia's internals, so there may be some glaring error.

### CODE REVIEW

**Comment by martingust** [comment]
@lsesol Sorry for late response to on this. I've just brought down the code it works great from what I can see. Would you mind sync this with master and I'll merge this in?

**Comment by lsesol** [comment]
@martingust Should be up to date. Cheers.

**Comment by martingust** [comment]
@lsesol It looks like you got some additional files in there as well, dist, config.js and package.json. Please remove those from the PR. Thanks!

File: dist/aurelia-ui-virtualization.js [hunk 1]
```diff
@@ -195,30 +195,51 @@ export class ArrayVirtualRepeatStrategy extends ArrayRepeatStrategy {
...
-    for (let i = 0, ii = splices.length; i < ii; ++i) {
-      let splice = splices[i];
-      let removed = splice.removed;
-      let removedLength = removed.length;
-      for (let j = 0, jj = removedLength; j < jj; ++j) {
-        let viewOrPromise = this._removeViewAt(repeat, splice.index + removeDelta + rmPromises.length, true, j, removedLength);
-        if (viewOrPromise instanceof Promise) {
-          rmPromises.push(viewOrPromise);
+    // do all splices replace existing entries?
+    let allSplicesAreInplace = splices.reduce( (prev,splice) => {
```
Comment (martingust) [inline dist/aurelia-ui-virtualization.js]: Please replace this with a for-loop for better performance

File: dist/aurelia-ui-virtualization.js [hunk 2]
```diff
@@ -195,30 +195,51 @@ export class ArrayVirtualRepeatStrategy extends ArrayRepeatStrategy {
...
-        if (viewOrPromise instanceof Promise) {
-          rmPromises.push(viewOrPromise);
+    // do all splices replace existing entries?
+    let allSplicesAreInplace = splices.reduce( (prev,splice) => { 
+      return prev && splice.removed.length===splice.addedCount; 
+    }, true);
+
+    // if so, optimise by just replacing affected visible views
+    if( allSplicesAreInplace ) {
+      splices.forEach( (splice) => {
```
Comment (martingust) [inline dist/aurelia-ui-virtualization.js]: Same here, please use for-loop

File: src/array-virtual-repeat-strategy.js [hunk 3]
```diff
@@ -101,30 +101,51 @@ export class ArrayVirtualRepeatStrategy extends ArrayRepeatStrategy {
...
-    for (let i = 0, ii = splices.length; i < ii; ++i) {
-      let splice = splices[i];
-      let removed = splice.removed;
-      let removedLength = removed.length;
-      for (let j = 0, jj = removedLength; j < jj; ++j) {
-        let viewOrPromise = this._removeViewAt(repeat, splice.index + removeDelta + rmPromises.length, true, j, removedLength);
-        if (viewOrPromise instanceof Promise) {
-          rmPromises.push(viewOrPromise);
+    // do all splices replace existing entries?
+    let allSplicesAreInplace = splices.reduce( (prev,splice) => {
```
Comment (martingust) [inline src/array-virtual-repeat-strategy.js:106]: Please replace reduce with a for-loop for better performance

File: src/array-virtual-repeat-strategy.js [hunk 4]
```diff
@@ -101,30 +101,51 @@ export class ArrayVirtualRepeatStrategy extends ArrayRepeatStrategy {
...
-        if (viewOrPromise instanceof Promise) {
-          rmPromises.push(viewOrPromise);
+    // do all splices replace existing entries?
+    let allSplicesAreInplace = splices.reduce( (prev,splice) => { 
+      return prev && splice.removed.length===splice.addedCount; 
+    }, true);
+
+    // if so, optimise by just replacing affected visible views
+    if( allSplicesAreInplace ) {
+      splices.forEach( (splice) => {
```
Comment (martingust) [inline src/array-virtual-repeat-strategy.js:112]: Same here, please use for-loop

**Comment by lsesol** [comment]
Apologies - I introduced that running the tests. I thought it would be cleaner to just start a new PR, so I'll close this one. Please see the new single file PR including the requested changes.

**Pull request closed without merging**